dmff/mbar.py

```diff
@@ -0,0 +1,194 @@
+        cm = 1. / (jax.numpy.exp(delta_u) * jax.numpy.array(self._nk).reshape(
+            (-1, 1))).sum(axis=0)
+        weight = cm / cm.sum()
+        if return_freeE:
+            f_new = -jax.numpy.log(jax.numpy.sum(jax.numpy.exp(cm)))
+            return weight, f_new
+        return weight
+
+    def _estimate_weight_numpy(self, unew_npy, return_cn=False):
+        unew_mean = unew_npy.mean()
+        du_1 = self._mbar.f_k.reshape((-1, 1)) - self._umat
+        delta_u = du_1 + unew_npy.reshape((1, -1)) - unew_mean - du_1.mean()
+        cn = 1. / (np.exp(delta_u) * self._nk.reshape((-1, 1))).sum(axis=0)
+        weight = cn / cn.sum()
+        if return_cn:
+            return weight, cn
+        else:
+            return weight
+
+    def _computeCovar(self, W, N_k):
+        K, N = W.shape
+        Ndiag = np.diag(N_k)
+        I = np.identity(K, dtype=np.float64)
+
+        S2, V = np.linalg.eigh(W @ W.T)
+        S2[np.where(S2 < 0.0)] = 0.0
+        Sigma = np.diag(np.sqrt(S2))
+
+        # Compute covariance
+        Theta = (V @ Sigma @ np.linalg.pinv(
+            I - Sigma @ V.T @ Ndiag @ V @ Sigma, rcond=1e-10) @ Sigma @ V.T)
+        return Theta
+
+    def compute_covar_mat(self, unew):
+        wnew = self._estimate_weight_numpy(unew)
+        wappend = np.concatenate(
+            [self._mbar.W_nk.T, wnew.reshape((1, -1))], axis=0)
+
+        N_k = np.zeros((self._mbar.N_k.shape[0] + 1, ))
+        N_k[:-1] = self._mbar.N_k[:]
+        newcov = self._computeCovar(wappend, N_k)
+        return newcov
+
+    def compute_variance(self, unew, prop):
+        wnew, cn = self._estimate_weight_numpy(unew, return_cn=True)
+        ca = (1. / cn).sum()
+        A_ave = (prop * wnew).sum()
+        W_A = wnew * (prop / A_ave)
+
+        wappend = np.concatenate(
+            [self._mbar.W_nk.T,
+             wnew.reshape((1, -1)),
+             W_A.reshape((1, -1))],
+            axis=0)
+
+        N_k = np.zeros((self._mbar.N_k.shape[0] + 2, ))
+        N_k[:-2] = self._mbar.N_k[:]
+
+        newcov = self._computeCovar(wappend, N_k)
+        return A_ave * A_ave * (newcov[-2, -2] + newcov[-1, -1] -
+                                2. * newcov[-1, -2])
+
+    def estimate_effective_sample(self, unew, decompose=False):
+        wnew, cn = self._estimate_weight_numpy(unew, return_cn=True)
+        eff_samples = 1. / (wnew**2).sum()
+        if decompose:
+            state_effect = np.zeros((len(self.states)))
+            argsort = np.argsort(wnew)[::-1][:int(eff_samples) + 1]
+            for nstate in range(state_effect.shape[0]):
+                istart = self._nk[:nstate].sum()
+                iend = istart + self._nk[nstate]
+                state_effect[nstate] = ((argsort > istart) &
+                                        (argsort < iend)).sum()
+            return eff_samples, state_effect
+        return eff_samples
+
+    def estimate_free_energy(self, unew):
```

Comment:
1. We should consider computing free energies between two target states.
2. Target states and params should be passed in as argument.